keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1838 @@
+   Git pre-submit (there are tools being written now that will also handle
+   x-kubernetes-validations)
+
+#### User wishes to use Kubebuilder to define a custom resource that embeds PodTemplate
+
+1. Go struct that declares custom resource references the go struct of v1.PodTemplate.
+2. Kubebuilder recognizes the Go IDL tags introduced by this KEP (Note that
+   Kubebuidler already has IDL tags:
+   https://book.kubebuilder.io/reference/markers/crd-validation.html, but they
+   are different than what we propose here)
+3. Kubebuilder generates appropriate OpenAPI for the CRD resulting in full
+   validation of the PodTemplate
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Risk: Migration to Declarative Validation introduce breaking change to API validation
+##### Mitigation: Ensure Invalid Objects Still Invalid
+
+Objects which previously did not pass validation must still not pass validation.
+This can be restated simply as: all currently handwritten validations must be
+included in the schema. If all validations are in the schema, it follows that the
+schema will reject the same set of objects as before.
+
+We can mitigate this by:
+
+1. All Validation Tests Have 100% Coverage
+2. All Validation Tests Succeed against native and schema-based validation backends
+
+To accomplish this, we will:
+
+1. New Presubmit Verify Script to Ensure 100% for future tests
+2. A General Validation Test Framework - Most tests are already written declaratively,
+we will refactor them to all use a general framework for validation tests. The
+framework will evaluate the validation test case first on native types, then convert
+the case to each supported version of the kind to ensure the error arises also
+from schema-based validation. The framework will have a way to make exceptions 
+for certain errors which are version-specific.
+3. Refactor all `validation_tests.go` unit tests to have 100% coverage of their 
+respective `validation.go` files as well as make use of the validation test framework.
+
+##### Mitigation: Ensure Valid Old Objects Still Valid
+
+Objects which previously passed validation must still pass validation. For our
+purposes this can be thought of as ensuring the schema-based validation does not
+raise new errors which previously did not exist.
+
+It is not possible to be 100% sure of this property, in fact, there may be
+new validations added to patch bugs in existing ones. But we can mitigate this 
+risk:
+
+1. Catch and fix any significant differences in schema validations
+  
+  During alpha and beta phases, we will run both validation systems in parallel
+  and return the errors from the old system to the user. We will report in logs
```

Comment:
That's interesting, so the new validation system basically will be a no-opt even during beta? Isn't that a deviation from beta features being enabled by default?

Reply:
Reflecting on it, even the alpha semantics are kinda weird. Enabling this feature doesn't actually enable this feature. It just enables the comparative checking of this feature to the existing one. Is that right?

Reply:
> beta features being enabled by default?

I think since https://github.com/kubernetes/enhancements/issues/3136 beta features are not enabled by default, unless you meant in another sense?

> That's interesting, so the new validation system basically will be a no-opt even during beta?
In beta we will begin to admit using declarative validation in resources for which it is enabled. 

The current proposal in the KEP is:

- Alpha: Runs both systems, logs differences, admits based off of native validation
- Beta: Runs both systems, logs differences, admits based off of declarative validation
- GA: Admits based of declarative validation

I agree it is weird. The main reason it ended up like this was to avoid writing invalid resources based on unfinished declarative validations, so the feature can be safely disabled, rolled back, etc. during development.

Very open to suggestions if you have alternative ideas for a cleaner approach

Reply:
> I think since https://github.com/kubernetes/enhancements/issues/3136 beta features are not enabled by default, unless you meant in another sense?

beta APIs are different from beta features. beta features are still generally enabled by default.

Reply:
> Alpha: Runs both systems, logs differences, admits based off of native validation

This semantic is a little strange is what I'm saying. Enabling the feature doesn't actually enable the feature.

Reply:
It may make sense to have two alpha stages, because the first alpha stage doesn't really permit truly enabling the feature. I would want at least one alpha stage where enabling the feature does truly enable it.

Reply:
Sounds okay to me.

Reply:
updated KEP